pkg/il/builder.go

```diff
@@ -0,0 +1,281 @@
+func (f *Builder) ResolveDouble(n string) {
+	f.op1(ResolveD, f.id(n))
+}
+
+// TResolveDouble appends the "tresolve_d" instruction to the byte code.
+func (f *Builder) TResolveDouble(n string) {
+	f.op1(TResolveD, f.id(n))
+}
+
+// ResolveRecord appends the "resolve_r" instruction to the byte code.
+func (f *Builder) ResolveRecord(n string) {
+	f.op1(ResolveM, f.id(n))
+}
+
+// TResolveRecord appends the "tresolve_r" instruction to the byte code.
+func (f *Builder) TResolveRecord(n string) {
+	f.op1(TResolveM, f.id(n))
+}
+
+// IPushBool appends the "ipush_b" instruction to the byte code.
+func (f *Builder) IPushBool(b bool) {
+	f.op1(APushB, boolToOpcode(b))
+}
+
+// IPushStr appends the "ipush_s" instruction to the byte code.
+func (f *Builder) IPushStr(s string) {
+	f.op1(APushS, f.id(s))
+}
+
+// IPushInt appends the "ipush_i" instruction to the byte code.
+func (f *Builder) IPushInt(i int64) {
+	a1, a2 := integerToOpcode(i)
+	f.op2(APushI, a1, a2)
+}
+
+// IPushDouble appends the "ipush_d" instruction to the byte code.
+func (f *Builder) IPushDouble(n float64) {
+	a1, a2 := doubleToOpcode(n)
+	f.op2(APushD, a1, a2)
+}
+
+// Xor appends the "xor" instruction to the byte code.
+func (f *Builder) Xor() {
+	f.op0(Xor)
+}
+
+// EQString appends the "eq_s" instruction to the byte code.
+func (f *Builder) EQString() {
+	f.op0(EqS)
+}
+
+// IEQString appends the "ieq_s" instruction to the byte code.
+func (f *Builder) IEQString(v string) {
+	f.op1(AEqS, f.id(v))
+}
+
+// EQBool appends the "eq_b" instruction to the byte code.
+func (f *Builder) EQBool() {
+	f.op0(EqB)
+}
+
+// IEQBool appends the "ieq_b" instruction to the byte code.
+func (f *Builder) IEQBool(v bool) {
+	f.op1(AEqB, boolToOpcode(v))
+}
+
+// EQInteger appends the "eq_i" instruction to the byte code.
+func (f *Builder) EQInteger() {
+	f.op0(EqI)
+}
+
+// IEQInteger appends the "eq_i" instruction to the byte code.
+func (f *Builder) IEQInteger(v int64) {
+	a1, a2 := integerToOpcode(v)
+	f.op2(AEqI, a1, a2)
+}
+
+// EQDouble appends the "eq_d" instruction to the byte code.
+func (f *Builder) EQDouble() {
+	f.op0(EqD)
+}
+
+// IEQDouble appends the "ieq_d" instruction to the byte code.
+func (f *Builder) IEQDouble(v float64) {
+	a1, a2 := doubleToOpcode(v)
+	f.op2(AEqD, a1, a2)
+}
+
+// Not appends the "not" instruction to the byte code.
+func (f *Builder) Not() {
+	f.op0(Not)
+}
+
+// Or appends the "or" instruction to the byte code.
+func (f *Builder) Or() {
+	f.op0(Or)
+}
+
+// And appends the "and" instruction to the byte code.
+func (f *Builder) And() {
+	f.op0(And)
+}
+
+// Lookup appends the "lookup" instruction to the byte code.
+func (f *Builder) Lookup() {
+	f.op0(Lookup)
+}
+
+// ILookup appends the "ilookup" instruction to the byte code.
+func (f *Builder) ILookup(v string) {
+	f.op1(ALookup, f.id(v))
+}
+
+// AllocateLabel allocates a new label value for use within the code.
+func (f *Builder) AllocateLabel() string {
+	l := fmt.Sprintf("L%d", len(f.labels))
```

Comment:
It is a convenience method for generating labels. You can use your own labels as well.

During encoding address based opcodes, you will either need to put the label first, then the jump instruction or vice-versa.

If you need to put the label first, then you can simply call SetLabelPos. Then, when you call jmp, it will use the label's address.
If you put the jmp first, then it will register and a fixup will be noted. When you actually call SetLabelPos, then the address will be recorded and any previous fixups are updated.
Both require you to bring your own label. AllocateLabel is a convenience way to come up with those labels in a stable way.